nifi-framework-bundle/nifi-framework-extensions/nifi-py4j-framework-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/componentstate.py

```diff
@@ -0,0 +1,171 @@
+#  Licensed to the Apache Software Foundation (ASF) under one or more
+#  contributor license agreements.  See the NOTICE file distributed with
+#  this work for additional information regarding copyright ownership.
+#  The ASF licenses this file to You under the Apache License, Version 2.0
+#  (the "License"); you may not use this file except in compliance with
+#  the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from enum import Enum
+from nifiapi.__jvm__ import JvmHolder
+from py4j.protocol import Py4JJavaError
+
+
+CLUSTER_SCOPE = JvmHolder.jvm.org.apache.nifi.components.state.Scope.CLUSTER
+LOCAL_SCOPE = JvmHolder.jvm.org.apache.nifi.components.state.Scope.LOCAL
+
+
+def convert_pyscope_to_jscope(scope):
+    """Converts Python Scope to Java Scope."""
+    if scope == Scope.LOCAL:
+        return LOCAL_SCOPE
+    else:
+        return CLUSTER_SCOPE
+
+def convert_pydict_to_jmap(dictionary):
+    """Converts a Python Dictionary to a Java Map"""
+    java_map = JvmHolder.jvm.java.util.HashMap()
+    for key, value in dictionary.items():
+        java_map.put(key, value)
+    return java_map
+
+class Scope(Enum):
+    CLUSTER = 1
+    LOCAL = 2
+
+
+class StateManager:
+    """
+    Python class wrapping the StateManager Java implementation.
+
+    The constructor receives a Java object implementing the Java StateManager interface.
+    The methods in Java StateManager have corresponding methods in this class.
+    Each method in this class propagates the method call to the Java StateManager object.
+    Each method may raise StateException in case the Java object fails to execute the particular operation.
+    """
+
+    _java_state_manager = None
+
+    def __init__(self, java_state_manager):
+        """
+        :param java_state_manager: Java StateManager object that performs all the StateManager-related operations.
+        """
+        self._java_state_manager = java_state_manager
+
+    def setState(self, state, scope):
+        """
+        :param state: (dict) Python dictionary containing the state to be saved.
+        :param scope: Scope.CLUSTER or Scope.LOCAL
+        """
+        try:
+            self._java_state_manager.setState(convert_pydict_to_jmap(state), convert_pyscope_to_jscope(scope))
+        except Py4JJavaError as exception:
+            raise StateException(str(exception)) from None
```

Comment:
It seems like it would be better to chain the original exception, as opposed to stringifying it as shown. Is there a particular reason for not chaining the exception?
```suggestion
            raise StateException("Set state failed") from exception
```

Reply:
I wanted to hide the Py4JJavaError instances, thought it would be too technical and might confuse ppl.